lib/scripting/objects/pinball-backend/emulator-state.ts

```diff
@@ -0,0 +1,173 @@
+	}
+
+	public updateState(state: WpcEmuWebWorkerApi.EmuStateAsic) {
+		if (state.wpc.lampState) {
+			this.currentLampState = this.normalizeValue(state.wpc.lampState);
+		}
+		if (state.wpc.solenoidState) {
+			//TODO unclear if we need to normalize
+			this.currentSolenoidState = state.wpc.solenoidState;
+		}
+		if (state.wpc.generalIlluminationState) {
+			//TODO unclear if we need to normalize
+			this.currentGIState = state.wpc.generalIlluminationState;
+		}
+		if (state.dmd.dmdShadedBuffer) {
+			this.dmdScreen = state.dmd.dmdShadedBuffer;
+		}
+		if (state.wpc.inputSwitchMatrixActiveColumn) {
+			this.switchState = state.wpc.inputSwitchMatrixActiveColumn;
+		}
+	}
+
+	public getSwitchState(index: number): number {
+		const matrixIndex: number = mapIndexToMatrixIndex(index);
+		return this.switchState[matrixIndex] || 0;
+	}
+
+	public getLampState(index: number): number {
+		const matrixIndex: number = mapIndexToMatrixIndex(index);
+		return this.currentLampState[matrixIndex] || 0;
+	}
+
+	public getSolenoidState(index: number): number {
+		const matrixIndex: number = index++;
+		return this.currentSolenoidState[matrixIndex] || 0;
+	}
+
+	public getGIState(index: number): number {
+		const matrixIndex: number = index++;
+		return this.currentGIState[matrixIndex] || 0;
+	}
+
+	/**
+	 * return changed lamps, index starts at 11..18, 21..28.. up to index 88
+	 */
+	public getChangedLamps(): number[][] {
+		const result: number[][] = this.getArrayDiff(this.lastSentLampState, this.currentLampState, mapIndexToMatrixIndex);
+		this.lastSentLampState = this.currentLampState;
+		return result;
+	}
+
+	/**
+	 * return changed solenoids, index starts at
+	 */
+	public getChangedSolenoids(): number[][] {
+		const result: number[][] = this.getArrayDiff(this.lastSentSolenoidState, this.currentSolenoidState, mapIndexToOneBasedIndex);
+		this.lastSentSolenoidState = this.currentSolenoidState;
+		return result;
+	}
+
+	public getChangedGI(): number[][] {
+		const result: number[][] = this.getArrayDiff(this.lastSentGIState, this.currentGIState, mapIndexToOneBasedIndex);
+		this.lastSentGIState = this.currentGIState;
+		return result;
+	}
+
+	// NOT IMPLEMENTED YET - needed for alphanumeric games only!
+	public ChangedLEDs(): number[][] {
+		return [];
+	}
+
+	public getDmdScreen(): Uint8Array {
+		return this.dmdScreen;
+	}
+
+	/**
+	 * map uint8 values to 0 or 1 (VisualPinball engine)
+	 */
+	private normalizeValue(input: Uint8Array): Uint8Array {
+		return input.map((value) => value > 127 ? 1 : 0);
+	}
+
+	/**
+	 * diff between two arrays equally sized arrays
+	 * returns 2 dimensional array with the result, eg [0, 5], [4, 44] -> means entry at offset 0 changed to 5, entry at offset 4 changed to 44
+	 */
+	private getArrayDiff(lastState: Uint8Array, newState: Uint8Array, offsetMapperFunction: (index: number) => number): number[][] {
+		const result: number[][] = [];
+		if (arraysEqual(lastState, newState)) {
+			return result;
+		}
```

Comment:
it would bail out as soon as a diff is found - but you're right, its more a improvement which might not bee needed - lets fix perf issues where they happen ;)
less code!